commonmark/src/main/java/org/commonmark/internal/inline/Scanner.java

```diff
@@ -0,0 +1,187 @@
+package org.commonmark.internal.inline;
+
+import org.commonmark.internal.util.CharMatcher;
+
+import java.util.List;
+
+public class Scanner {
+
+    // Lines without newlines at the end. The scanner will yield `\n` between lines because they're significant for
+    // parsing and the final output. There is no `\n` after the last line.
+    private final List<CharSequence> lines;
+    // Which line we're at.
+    private int lineIndex;
+    // The index within the line. If index == length(), we pretend that there's a `\n` and only advance after we yield
+    // that.
+    private int index;
+
+    // Current line or "" if at the end of the lines (using "" instead of null saves a null check)
+    private CharSequence line = "";
+    private int lineLength = 0;
+
+    // TODO: Visibility
+    public Scanner(List<CharSequence> lines, int lineIndex, int index) {
+        this.lines = lines;
+        this.lineIndex = lineIndex;
+        this.index = index;
+        if (!lines.isEmpty()) {
+            line = lines.get(lineIndex);
+            lineLength = line.length();
+        }
+    }
+
+    public char peek() {
+        if (index < lineLength) {
+            return line.charAt(index);
+        } else {
+            if (lineIndex < lines.size() - 1) {
+                return '\n';
+            } else {
+                // Don't return newline for end of last line
+                return '\0';
```

Comment:
I can see we using this symbol in multiple places. What do you think to extract and also use the constant from Character class? `Character.MIN_VALUE`?

Reply:
Good idea, yeah! `Character.MIN_VALUE` would work, but I think something even more specific would be nicer like `Parsing.EOF` or `Parsing.END`.